test/metabase/db/custom_migrations_test.clj

```diff
@@ -2346,3 +2369,143 @@
                    (no-stage-pattern multi-stage-question-id)
                    (no-stage-pattern multi-stage-model-id)]
                   (query-parameter-mappings))))))))
+
+(deftest set-stage-number-in-viz-settings-parameter-mappings-test
+  (testing "v52.2024-11-12T15:13:18"
+    (impl/test-migrations ["v52.2024-11-12T15:13:18"] [migrate!]
+      (let [{:keys [dashboard-id
+                    single-stage-question-id
+                    native-question-id
+                    multi-stage-question-id
+                    multi-stage-model-id]}
+            (insert-stage-number-data!)
+
+            viz-settings-generator
+            (fn [dimensions]             ; 4 dimensions are consumed
+              ;; crazy as it sounds, the FE uses the JSON encoded form of the dimensions as IDs and keys
+              (let [dimension-strs (mapv json/generate-string dimensions)
+                    dimension-keys (mapv keyword dimension-strs)]
+                (fn [[target-id0 target-id1 target-id2]]
```

Comment:
This inner function is really hard to follow, mostly because of the vertical length. Can we either abstract an inner chunk into a separate function (perhaps from index to `:parameterMapping` map?), turn it into a `for` or similar, or use the horizontal space a bit more effectively?